pkg/utils/sort_test.go

```diff
@@ -112,19 +112,41 @@ func Test_SortEnvVarSlice(t *testing.T) {
 				},
 			},
 			expectEnvVar: []corev1.EnvVar{
+				{
+					Name:  "ENV_B",
+					Value: "valueB",
+				},
 				{
 					Name:  "ENV_A",
-					Value: "$(ENV_B)_suffix",
+					Value: "$$(ENV_C)_$(ENV_B)_suffix",
 				},
 				{
 					Name:  "ENV_C",
-					Value: "$$(ENV_A)_$(ENV_B)_suffix",
+					Value: "$(ENV_B)_suffix",
+				},
+			},
+		},
+		"env-dependencies-unescaped": {
+			envVar: []corev1.EnvVar{
+				{
+					Name:  "ENV_A",
+					Value: "$$$(ENV_B)_suffix",
 				},
 				{
 					Name:  "ENV_B",
 					Value: "valueB",
 				},
 			},
+			expectEnvVar: []corev1.EnvVar{
+				{
+					Name:  "ENV_B",
+					Value: "valueB",
+				},
+				{
+					Name:  "ENV_A",
+					Value: "$$$(ENV_B)_suffix",
+				},
+			},
 		},
 		"env-fake-transitive": {
 			envVar: []corev1.EnvVar{
```

Comment:
I think the reference in the input string would be unchanged if a variable could not be resolved. Sorting the environment variables that contain unchanged references will change the meaning of value strings. They were semantically incorrect, but grammatically correct.

Reply:
It seems these cases are rare in normal, and it makes the final order hard to predict if we do so. Currently, I think we should just make it semantically correct to reduce the mental burden. If one wants a raw `$(...)`, he'll need to escape the expression with `$$(...)` explicitly.